### PR TITLE
Snap selectors

### PR DESCRIPTION
**Pull request by ericpre**
## PR Summary

Snap SpanSelector position to specific values.

```python

import matplotlib.pyplot as plt
from matplotlib.widgets import SpanSelector, RectangleSelector
import numpy as np

y = np.arange(25)
snap_values = np.append(y-0.5, [y[-1] + 0.5])
snap_values_2D = np.array([y]*2)

fig, ax = plt.subplots()
ax.plot(y, 'o')

span = SpanSelector(ax, print, direction='horizontal',
                    interactive=True,
                    drag_from_anywhere=True,
                    snap_values=snap_values)
```

### CODE REVIEW

**Comment by timhoffm** [comment]
What is the use case for this? I could imagine snapping to artist bounding boxes or points in a `Line2D`, but snapping to user specified values seems not too useful (or a reasonable user-facing API). Why and how do you anticipate this will be used? Without a clear idea of that, I'm afraid this complicates the code base and adds a maintainance burden without a clear benefit.

**Comment by timhoffm** [comment]
Ok, for a SpanSelector this makes sense.

I'm much more unclear what the API and desired effects should be for the 2D selectors. It seems that there x and y are just independent sets of snap-positions per axis. 1) You might want to have correlations, e.g. for `scatter(x, y)` with random x, y I could want to snap a `PolygonSelector` to data points only. 2) If they are supposed to be independent a 2D array is not a good choice. I might have a different number of snap positions for x and y.

IMHO one would have to carefully think through use cases and suitable API & behavior per Selector separately. Otherwise, we may end in a corner we don't want to be in. How about starting with SpanSelector, which seems pretty straight forward? And leave the other selectors for later?

On a more conecptual note, it may (maybe at a later time) be desirable to inject the snap logic, i.e. pass a function in that does the snapping: `snap(x, y) -> x_snapped, y_snapped`. We should keep this in mind when naming: `snap_values=snap` is okish but maybe there's a better name. We could also consider building the internal logic around such an approach.

**Comment by ericpre** [comment]
For 2D selectors, it will be the same for images, which have been acquired with a digital camera, where the intensity of each pixels is expected to measure a number of events hitting this specific pixels.
```python
import matplotlib.pyplot as plt
from matplotlib.widgets import RectangleSelector
import numpy as np

values = np.random.random(size=(8, 8))
snap_values_2D = np.tile(np.mgrid[-0.5:8.5], (2, 1))

_, ax = plt.subplots()
plt.imshow(values)

rect = RectangleSelector(ax, print,
                         interactive=True,
                         drag_from_anywhere=True,
                         snap_values=snap_values_2D)
```

For scatter, I don't know and indeed I don't think that it makes sense.

> On a more conecptual note, it may (maybe at a later time) be desirable to inject the snap logic, i.e. pass a function in that does the snapping: `snap(x, y) -> x_snapped, y_snapped`. We should keep this in mind when naming: `snap_values=snap` is okish but maybe there's a better name. We could also consider building the internal logic around such an approach.

I explore a bit this option, but I couldn't get anyway nice. The approach of this PR has the big advantage of being simple.

File: lib/matplotlib/widgets.py [hunk 1]
```diff
@@ -2238,6 +2238,10 @@ def on_select(min: float, max: float) -> Any
         If `True`, the event triggered outside the span selector will be
         ignored.
 
+    snap_values : 1D array-like, default: None
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:2241]:
```suggestion
    snap_values : 1D array-like, optional
```
We write it this way if `None` is only the placeholder for "not given" (c.f. https://matplotlib.org/devdocs/devel/documenting_mpl.html#default-values)

File: lib/matplotlib/widgets.py [hunk 2]
```diff
@@ -2238,6 +2238,10 @@ def on_select(min: float, max: float) -> Any
...
+        Snap the extents of the selector to the values defined in
+        ``snap_values``.
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:2243]:
```suggestion
        Snap the selector edges to the given values.
```

File: lib/matplotlib/widgets.py [hunk 3]
```diff
@@ -2577,6 +2582,16 @@ def _contains(self, event):
...
 
+    @staticmethod
+    def _snap(values, snap_values):
+        """Snap values to a given array values (snap_values)."""
+        indices = np.empty_like(values, dtype="uint")
+        # take into account machine precision
+        eps = np.min(np.abs(np.diff(snap_values))) * 1e-12
+        for i, v in enumerate(values):
+            indices[i] = np.abs(snap_values - v + np.sign(v) * eps).argmin()
+        return snap_values[indices]
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:2593]:
```suggestion
        indices = np.empty_like(values, dtype="uint")
        # take into account machine precision
        eps = np.min(np.abs(np.diff(snap_values))) * 1e-12
        return tuple(
            snap_values[np.abs(snap_values - v + np.sign(v) * eps).argmin()]
            for v in values)
```
is maybe a bit easier as you don't have to carry the indices.

File: lib/matplotlib/widgets.py [hunk 4]
```diff
@@ -2857,6 +2874,7 @@ def onselect(eclick: MouseEvent, erelease: MouseEvent)
     use_data_coordinates : bool, default: False
         If `True`, the "square" shape of the selector is defined in
         data coordinates instead of display coordinates.
+
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:2877]:
```suggestion
```
Unneeded empty line.

**Review by timhoffm — APPROVED** [review]

**Review by oscargus — APPROVED** [review]

**Pull request merged**